File: Algorithms/C++/00234-Palindrome Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        ListNode* temp = head;
        int size{0};
        while(temp != nullptr){
            size++;
            temp = temp->next;
        }
        
        // get head of 2nd half of list
        int half = (size/2);
        ListNode* secondHead = head;
        while(half > 0){
            secondHead = secondHead->next;
            half--;
        }
        // if list size is odd, move up one node for secondHead
        if(size % 2 == 1) secondHead = secondHead->next;
        
        // reverse 1st half of list
        half = (size/2);
        temp = head;
        ListNode* prev = nullptr;
        while(half > 0){
            ListNode* next = temp->next;
            temp->next = prev;
            prev = temp;
            temp = next;
            half--;
        }

        // iterate through both second halves of list 
        half = (size/2);
        while(half > 0){
            if(prev->val != secondHead->val){
                std::cout << prev->val << " " << secondHead->val << "\n";
                return false;
            }
            prev = prev->next;
            secondHead = secondHead->next;
            half--;
        }

        return true;
    }
};

/*
    O(n) and O(n)
    - go through list, put in array, and check via array
    
    O(n) and O(1)
    1 - 2 | 2 - 1

    - check the size -> even or odd
    - iterate through half of list (size/2)
    - store pointer at half+1
    - reverse 1st half of list
    - iterate through 1st half and 2nd half to check if values are equal

*/

File: Algorithms/Python3/00234-Palindrome Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        dummy = head
        size = 0
        while dummy:
            size += 1
            dummy = dummy.next
        
        midpoint = size//2
        curr = head
        prev = None
        # reverse left side of list
        while midpoint > 0:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
            midpoint -= 1
        # case where size is odd
        if size % 2 == 1:
            curr = curr.next

        # check left and right lists for equality
        # right pointer = curr | left pointer = prev
        while curr and prev:
            if curr.val != prev.val:
                return False
            curr = curr.next
            prev = prev.next
        
        return True
        
'''
    tc: O(n)
    sc: O(1)

    notice that left half of list is a reversed right half
    
    - find length of list by going through once
    - divide by 2 = midpoint
    - reverse left half of list UP TO midpoint
    - iterate through both lists to check if they are the same
    

1 - 2 - 2 - 1

1 - 2 - 3 - 2 - 1
2 - 1 - 3 - 2 - 1




'''